content/issue-2/optimize-rust-comptime.md

```diff
@@ -0,0 +1,199 @@
+
+So a few days ago, I decided to spend some time researching whether I could optimize the compilation speed a bit more.
+
+What shocked me was that there were some very simple methods that, with just a little effort, produced astonishing results. I feel like I can describe them as low-hanging fruits, silver bullets, or even free lunch 🤯.
+
+---
+
+P.S. I highly recommend [matklad](https://github.com/matklad)'s blog (He is the original author of IntelliJ Rust and rust-analyzer):
+- [Fast Rust Builds](https://matklad.github.io/2021/09/04/fast-rust-builds.html)
+- [Delete Cargo Integration Tests](https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html)
+
+Most of the methods I used are discussed here, and he explains them clearly. If not otherwise indicated, all quotes in this article come from here.
+
+Although there are quite some articles talking about how to optimize Rust compilation speed (e.g., [Tips for Faster Rust Compile Times](https://endler.dev/2020/rust-compile-times/)), I still want to write another one to share my step-by-step process. Each optimization point comes with a corresponding PR, and you can combine the [commit history](https://github.com/risingwavelabs/risingwave/commits/main?after=d8198fa138003e1f1431053f4f5f09e4a5fa8fd8+69&branch=main&qualified_name=refs%2Fheads%2Fmain) to see the effect of each optimization point by comparing the CI pages before and after its PR.
+
+---
+
+P.P.S. Results after optimization: main workflow is now 27 minutes at the fastest, and PR workflow is now 16 minutes at the fastest, with most taking around 17-19 minutes.
+
+# Valuable data and charts to find the bottlenecks
+
+> Build times are a fairly easy optimization problem: it’s trivial to get direct feedback (just time the build), there are a bunch of tools for profiling, and you don’t even need to come up with a representative benchmark.
+
+When trying to optimize anything, it would be good to have some profiling data and charts to find out the bottlenecks. Luckily, we do have some nice ones for optimizing CI time.
+
+## CI Waterfall & DAG Graph
+
+We use Buildkite for our CI, and the normal view of a page (such as [Build #17099](https://buildkite.com/risingwavelabs/pull-request/builds/17099)) looks like this:
+
+![buildkite-1.png](https://xxchan.github.io/assets/img/comptime/buildkite-1.png)
+
+Buildkite has two very useful hidden pages, located at `/waterfall` and `/dag`, respectively, which show:
+
+![buildkite-waterfall.png](https://xxchan.github.io/assets/img/comptime/buildkite-waterfall.png)
+
+![buildkite-dag.png](https://xxchan.github.io/assets/img/comptime/buildkite-dag.png)
+
+From the graphs, we can see clearly that the biggest bottleneck is in the path of simulation build -> recovery test.
```

Comment:
Can you help the reader quickly get why `the biggest bottleneck is in the path of simulation build -> recovery test`?